### ISSUE TITLE
Unable to use FD approximants EccentricFD and TaylorF2Ecc in pycbc Inference

### ISSUE THREAD

**Issue by CalNaTNS** [gwastro/pycbc#4966]
Hi everyone,
I recently encountered an issue while performing injection analyses with EccentricFD and TaylorF2Ecc. When I initiated the calculation, I immediately ran into an error saying that the approximant EccentricFD( and TaylorF2Ecc) was unavailable. 
I checked that EccentricFD is available in the current version of PyCBC I am using (using the function `get_fd_waveform()`), but from the error message, pycbc seems to attempt to call EccentricFD using `get_td_waveform()`. This behavior is confusing since EccentricFD is a frequency-domain approximant, not a time-domain one. Besides, I met no error when creating the injection data file `injection.hdf`
The following is the error message and my script:

```
Traceback (most recent call last):
  File "/home/calnaky/anaconda3/envs/bilby/bin/pycbc_inference", line 125, in <module>
    model = models.read_from_config(cp)
  File "/home/calnaky/anaconda3/envs/bilby/lib/python3.9/site-packages/pycbc/inference/models/__init__.py", line 186, in read_from_config
    return get_model(name).from_config(cp, **kwargs)
  File "/home/calnaky/anaconda3/envs/bilby/lib/python3.9/site-packages/pycbc/inference/models/gaussian_noise.py", line 589, in from_config
    strain_dict, psd_strain_dict = data_from_cli(opts, **data_args)
  File "/home/calnaky/anaconda3/envs/bilby/lib/python3.9/site-packages/pycbc/inference/models/data_utils.py", line 424, in data_from_cli
    strain_dict = strain_from_cli_multi_ifos(opts, instruments,
  File "/home/calnaky/anaconda3/envs/bilby/lib/python3.9/site-packages/pycbc/inference/models/data_utils.py", line 50, in strain_from_cli_multi_ifos
    return strain.from_cli_multi_ifos(*args, **kwargs)
  File "/home/calnaky/anaconda3/envs/bilby/lib/python3.9/site-packages/pycbc/strain/strain.py", line 447, in from_cli_multi_ifos
    strain[ifo] = from_cli_single_ifo(opt, ifo,
  File "/home/calnaky/anaconda3/envs/bilby/lib/python3.9/site-packages/pycbc/strain/strain.py", line 436, in from_cli_single_ifo
    return from_cli(single_det_opt,
  File "/home/calnaky/anaconda3/envs/bilby/lib/python3.9/site-packages/pycbc/strain/strain.py", line 317, in from_cli
    injector.apply(strain, opt.channel_name.split(':')[0],
  File "/home/calnaky/anaconda3/envs/bilby/lib/python3.9/site-packages/pycbc/inject/inject.py", line 572, in apply
    signal = self.make_strain_from_inj_object(inj, delta_t,
  File "/home/calnaky/anaconda3/envs/bilby/lib/python3.9/site-packages/pycbc/inject/inject.py", line 639, in make_strain_from_inj_object
    hp, hc = get_td_waveform(inj, delta_t=delta_t, f_lower=f_l,
  File "/home/calnaky/anaconda3/envs/bilby/lib/python3.9/site-packages/pycbc/waveform/waveform.py", line 591, in get_td_waveform
    raise ValueError("Approximant %s not available" %
ValueError: Approximant EccentricFD not available
```

```
[model]
name = gaussian_noise
low-frequency-cutoff = 10.0

[variable_params]
; waveform parameters that will vary in MCMC
delta_tc =
mass1 =
mass2 =
eccentricity =
distance =
coa_phase =
inclination =
polarization =
ra =
dec =
[static_params]
; waveform parameters that will not change in MCMC
approximant = EccentricFD
f_lower = 10.0
f_ref = 10.0
; we'll set the tc by using the trigger time in the data
; section of the config file + delta_tc
trigger_time = ${data|trigger-time}

[prior-delta_tc]
; coalescence time prior
name = uniform
min-delta_tc = -0.1
...

```

Thanks!

**Comment by ahnitz** [comment]
@CalNaTNS Feel free to reach out to me at my public email ahnitz (at) syr.edu if you'd like access to the slack. I know some others who have some experience doing eccentric PE who may be able to provide advice.

**Comment by CalNaTNS** [comment]
@ahnitz Thank you very much for your reply! Should I add the time estimator to the function `get_imr_duration()` or somewhere else? I previously used that function for my waveform plugin and I noticed that it requires the initial frequency, component masses, and component spins on the z-axis. Since EccentricFD is a no-spin waveform, how should I handle the spin parameters needed? Is it reasonable that I directly use the length estimator for TaylorF2 and set all the spins to zero?

Besides, I really appreciate your help and I would love to access to the slack. I will email you via my school email calnatns (at) whu.edu.cn.

Thanks again for your time and help!

**Comment by ahnitz** [comment]
@CalNaTNS You can either add a plugin function to return the expected signal duration. You are free to ignore arguments like spin if you want. It's just that the function should take arbitrary keyword arguments. So if someone passes a spin, no error should result; rather it can simply be not used. 

There's an example of generally writing plugins here 

https://pycbc.org/pycbc/latest/html/waveform_plugin.html

The main difference for the duration is that the setup.py (or pyproject.toml if you prefer those) should have the entry point name by pycbc.waveform.length rather than say pycbc.waveform.fd/td which would be for the waveform generation itself. 

Alternatively, you can provide a function to the dictionary here.
https://github.com/gwastro/pycbc/blob/master/pycbc/waveform/waveform.py#L1113

**Comment by CalNaTNS** [comment]
@ahnitz Thanks for your advice. They are useful! After adding the estimators for EccentricFD and TaylorF2Ecc to the dictionary [here,](https://github.com/gwastro/pycbc/blob/master/pycbc/waveform/waveform.py#L1113) I can run the programs now without any errors, and the results seem good. 
However, I noticed that when the distance is set too far away (e.g., 4000Mpc), the posteriors cannot converge, and the injected parameters cannot be recovered for both waveforms. While the distance is set to a smaller value (e.g., 300Mpc), the median value of the distance is still about 100 Mpc bigger than the injected value. Is that a normal case? Or did I set something incorrectly? I suppose it should not happen since some GW events are found to occur at a great distance away, like 4000+ Mpc. (e.g. GW190521)
